promql/engine.go

```diff
@@ -4147,14 +4148,50 @@ func NewHashRatioSampler() *HashRatioSampler {
 	return &HashRatioSampler{}
 }
 
-func (*HashRatioSampler) sampleOffset(_ int64, sample *Sample) float64 {
+func (s *HashRatioSampler) sampleOffset(_ int64, sample *Sample) float64 {
+	return s.SampleOffsetByLabels(sample.Metric)
+}
+
+// SampleOffsetByLabels returns a deterministic sampling offset in the range [0, 1)
+// derived from the hash of the provided metric labels.
+//
+// The offset is computed by normalizing the 64-bit hash value of the label set
+// to a float64 fraction of math.MaxUint64. This ensures that metrics with the
+// same label set always produce the same offset, while different label sets
+// produce uniformly distributed offsets suitable for sampling decisions.
+//
+// Note that this function allows for the actual labels.Labels to be passed in,
+// by passing the need to pass in a Sample. This allows downstream projects, such as
```

Comment:
[Nit] Assuming you mean bypassing :) A bit difficult to make sense of otherwise.
```suggestion
// bypassing the need to pass in a Sample. This allows downstream projects, such as
```